### PR TITLE
refactor(AS): refactor AS group resource method

### PR DESCRIPTION
**Pull request by deer-hang**
**What this PR does / why we need it**:

- The original method logic is difficult to understand and has hidden dangers. Therefore, the original method was modified and adjusted.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

- commit1: refactor method `checkASGroupInstancesInService`.
- commit2: refactor method `checkASGroupInstanceRemove`.
- commit3: refactor method `checkASGroupRemoved`.
- commit4: refactor method `resourceASGroupDelete`.

-> After deleting the cloud AS group and running `terraform plan`, the following prompt will appear:
image

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/as' TESTARGS='-run TestAccASGroup_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/as -v -run TestAccASGroup_basic -timeout 360m -parallel 4
=== RUN   TestAccASGroup_basic
=== PAUSE TestAccASGroup_basic
=== CONT  TestAccASGroup_basic
--- PASS: TestAccASGroup_basic (339.95s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/as        340.007s
```

```
make testacc TEST='./huaweicloud/services/acceptance/as' TESTARGS='-run TestAccASGroup_withEpsId'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/as -v -run TestAccASGroup_withEpsId -timeout 360m -parallel 4
=== RUN   TestAccASGroup_withEpsId
=== PAUSE TestAccASGroup_withEpsId
=== CONT  TestAccASGroup_withEpsId
--- PASS: TestAccASGroup_withEpsId (171.08s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/as        171.132s
```

```
make testacc TEST='./huaweicloud/services/acceptance/as' TESTARGS='-run TestAccASGroup_forceDelete'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/as -v -run TestAccASGroup_forceDelete -timeout 360m -parallel 4
=== RUN   TestAccASGroup_forceDelete
=== PAUSE TestAccASGroup_forceDelete
=== CONT  TestAccASGroup_forceDelete
--- PASS: TestAccASGroup_forceDelete (206.63s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/as        206.673s
```
```
make testacc TEST='./huaweicloud/services/acceptance/as' TESTARGS='-run TestAccASGroup_sourceDestCheck'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/as -v -run TestAccASGroup_sourceDestCheck -timeout 360m -parallel 4
=== RUN   TestAccASGroup_sourceDestCheck
=== PAUSE TestAccASGroup_sourceDestCheck
=== CONT  TestAccASGroup_sourceDestCheck
--- PASS: TestAccASGroup_sourceDestCheck (160.95s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/as        161.022s
```

### CODE REVIEW

File: huaweicloud/services/as/resource_huaweicloud_as_group.go [hunk 1]
```diff
@@ -768,66 +783,71 @@ func resourceASGroupUpdate(ctx context.Context, d *schema.ResourceData, meta int
 	return resourceASGroupRead(ctx, d, meta)
 }
 
+func forceDeleteASGroup(ctx context.Context, asClient *golangsdk.ServiceClient, d *schema.ResourceData) error {
+	if err := groups.ForceDelete(asClient, d.Id()).ExtractErr(); err != nil {
+		return fmt.Errorf("error deleting AS group %s: %s", d.Id(), err)
+	}
+
+	if err := checkASGroupRemoved(ctx, asClient, d.Id(), d.Timeout(schema.TimeoutDelete)); err != nil {
+		return fmt.Errorf("error deleting AS group %s: %s", d.Id(), err)
```
Comment (Lance52259) [inline huaweicloud/services/as/resource_huaweicloud_as_group.go:792]: seems this error message is incorrect and should be describe the waiting errors.

File: huaweicloud/services/as/resource_huaweicloud_as_group.go [hunk 2]
```diff
@@ -768,66 +783,71 @@ func resourceASGroupUpdate(ctx context.Context, d *schema.ResourceData, meta int
...
+	}
+	return nil
+}
+
+func deleteAllInstancesFromASGroup(ctx context.Context, asClient *golangsdk.ServiceClient, d *schema.ResourceData,
+	allIns []instances.Instance) error {
+	for _, ins := range allIns {
+		if ins.LifeCycleStatus != "INSERVICE" {
+			return fmt.Errorf("can't delete the AS group %s: some instances are not in INSERVICE but in %s, "+
```
Comment (Lance52259) [inline huaweicloud/services/as/resource_huaweicloud_as_group.go:801]:
```suggestion
			return fmt.Errorf("cannot delete the AS group %s: some instances are not in 'INSERVICE' status, actually is '%s', "+
```
`can't` is the colloquial expression, `cannot` uses here be better. other places ditto.

Comment (deer-hang) [reply huaweicloud/services/as/resource_huaweicloud_as_group.go:801]: I will check the error message in another `PR`.

File: huaweicloud/services/as/resource_huaweicloud_as_group.go (hunk 1 above)
Comment (deer-hang) [reply huaweicloud/services/as/resource_huaweicloud_as_group.go:792]: I will check the error message in another `PR`.

**Pull request merged**